File: examples/example_concurrent.c
/*
 * Sauron Library - Concurrent Stress Test
 *
 * Exercises thread safety by running simultaneous:
 * - Random IP readers (multiple threads)
 * - Bulk score updaters (multiple threads)
 * - Decay operations (single thread, periodic)
 *
 * Build:
 *   gcc -O2 -pthread -o example_concurrent example_concurrent.c \
 *       -I../include -L../src/.libs -lsauron -Wl,-rpath,../src/.libs
 *
 * Run:
 *   ./example_concurrent [duration_seconds] [num_reader_threads] [num_writer_threads]
 *   ./example_concurrent 10 4 2   # 10 seconds, 4 readers, 2 writers
 *
 * Copyright (c) 2024-2026, Ron Dilley
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <time.h>
#include <pthread.h>
#include <stdatomic.h>
#include <sched.h>
#include <locale.h>
#include <sauron.h>

/* Configuration */
#define DEFAULT_DURATION_SEC 10
#define DEFAULT_READER_THREADS 4
#define DEFAULT_WRITER_THREADS 2
#define DECAY_INTERVAL_MS 500
#define BULK_BATCH_SIZE 10000

/* Global shared state */
static sauron_ctx_t *g_ctx;
static atomic_int g_running;
static atomic_uint_least64_t g_total_reads;
static atomic_uint_least64_t g_total_writes;
static atomic_uint_least64_t g_total_decays;
static atomic_uint_least64_t g_read_errors;
static atomic_uint_least64_t g_write_errors;

/* Simple PRNG for thread-local randomness (xorshift32) */
static uint32_t xorshift32(uint32_t *state)
{
    uint32_t x = *state;
    x ^= x << 13;
    x ^= x >> 17;
    x ^= x << 5;
    *state = x;
    return x;
}

/* Generate a random non-bogon IP */
static uint32_t random_valid_ip(uint32_t *rng_state)
{
    uint32_t ip;
    int attempts = 0;

    ip = xorshift32(rng_state);
    /* Ensure we get a valid first octet (1-126, 128-223) */
    uint8_t first = (ip >> 24) & 0xFF;
    if (first == 0 || first == 127 || first >= 224) {
        /* Remap to valid range */
        first = 1 + (first % 126);  /* 1-126 */
        if (first >= 127) first++;   /* Skip 127 */
        ip = (ip & 0x00FFFFFF) | ((uint32_t)first << 24);
    }
    (void)attempts;  /* Unused after bogon check removal */

    return ip;
}

static double get_time(void)
{
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec / 1e9;
}

/*
 * Reader thread: Continuously reads random IPs
 */
static void *reader_thread(void *arg)
{
    (void)arg;
    uint32_t rng_state = (uint32_t)time(NULL) ^ (uint32_t)(uintptr_t)pthread_self();
    uint64_t local_reads = 0;

    while (atomic_load(&g_running)) {
        uint32_t ip = random_valid_ip(&rng_state);
        int16_t score = sauron_get_u32(g_ctx, ip);
        (void)score;  /* Used */

        local_reads++;

        /* Update global counter periodically */
        if ((local_reads & 0x3FFF) == 0) {
            atomic_fetch_add(&g_total_reads, local_reads);
            local_reads = 0;
            sched_yield();
        }
    }

    /* Final update */
    atomic_fetch_add(&g_total_reads, local_reads);
    return NULL;
}

/*
 * Writer thread: Continuously sets/increments random IPs
 */
static void *writer_thread(void *arg)
{
    (void)arg;
    uint32_t rng_state = (uint32_t)time(NULL) ^ (uint32_t)(uintptr_t)pthread_self() ^ 0xDEADBEEF;
    uint64_t local_writes = 0;

    while (atomic_load(&g_running)) {
        uint32_t ip = random_valid_ip(&rng_state);

        /* Mix of set and increment operations */
        int op = xorshift32(&rng_state) % 3;
        int16_t value = (int16_t)((xorshift32(&rng_state) % 1000) - 500);

        switch (op) {
        case 0:
            sauron_set_u32(g_ctx, ip, value);
            break;
        case 1:
            sauron_incr_u32(g_ctx, ip, value);
            break;
        case 2:
            sauron_decr_u32(g_ctx, ip, (int16_t)abs(value));
            break;
        }

        local_writes++;

        /* Update global counter periodically */
        if ((local_writes & 0xFFF) == 0) {
            atomic_fetch_add(&g_total_writes, local_writes);
            local_writes = 0;
            sched_yield();
        }
    }

    atomic_fetch_add(&g_total_writes, local_writes);
    return NULL;
}

/*
 * Bulk loader thread: Periodically bulk loads batches
 */
static void *bulk_loader_thread(void *arg)
{
    (void)arg;
    uint32_t rng_state = (uint32_t)time(NULL) ^ 0xCAFEBABE;
    char *buffer = malloc(BULK_BATCH_SIZE * 32);  /* ~32 bytes per line */

    if (!buffer) {
        fprintf(stderr, "ERROR: Failed to allocate bulk buffer\n");
        return NULL;
    }

    while (atomic_load(&g_running)) {
        /* Generate a batch of random updates */
        char *p = buffer;
        for (int i = 0; i < BULK_BATCH_SIZE; i++) {
            uint32_t ip = random_valid_ip(&rng_state);
            int16_t value = (int16_t)((xorshift32(&rng_state) % 200) - 100);
            int is_relative = xorshift32(&rng_state) % 2;

            int a = (ip >> 24) & 0xFF;
            int b = (ip >> 16) & 0xFF;
            int c = (ip >> 8) & 0xFF;
            int d = ip & 0xFF;

            if (is_relative) {
                p += sprintf(p, "%d.%d.%d.%d,+%d\n", a, b, c, d, value);
            } else {
                p += sprintf(p, "%d.%d.%d.%d,%d\n", a, b, c, d, value);
            }
        }

        sauron_bulk_result_t result;
        int ret = sauron_bulk_load_buffer(g_ctx, buffer, p - buffer, &result);

        if (ret == SAURON_OK) {
            atomic_fetch_add(&g_total_writes, result.sets + result.updates);
        }

        /* Small delay between bulk loads */
        usleep(10000);  /* 10ms */
    }

    free(buffer);
    return NULL;
}

/*
 * Decay thread: Periodically applies decay
 */
static void *decay_thread(void *arg)
{
    (void)arg;

    while (atomic_load(&g_running)) {
        uint64_t modified = sauron_decay(g_ctx, 0.99f, 1);
        atomic_fetch_add(&g_total_decays, modified);

        /* Sleep between decay cycles */
        usleep(DECAY_INTERVAL_MS * 1000);
    }

    return NULL;
}

/*
 * Stats reporter thread: Periodically prints statistics
 */
static void *stats_thread(void *arg)
{
    (void)arg;
    uint64_t last_reads = 0;
    uint64_t last_writes = 0;
    double last_time = get_time();
    int interval = 0;

    while (atomic_load(&g_running)) {
        sleep(1);
        interval++;

        uint64_t reads = atomic_load(&g_total_reads);
        uint64_t writes = atomic_load(&g_total_writes);
        double now = get_time();
        double elapsed = now - last_time;

        uint64_t read_delta = reads - last_reads;
        uint64_t write_delta = writes - last_writes;

        printf("  [%2ds] Reads: %'12lu (%'.1fM/s) | Writes: %'12lu (%'.1fM/s) | "
               "Count: %'lu | Blocks: %lu\n",
               interval,
               (unsigned long)reads, read_delta / elapsed / 1e6,
               (unsigned long)writes, write_delta / elapsed / 1e6,
               (unsigned long)sauron_count(g_ctx),
               (unsigned long)sauron_block_count(g_ctx));

        last_reads = reads;
        last_writes = writes;
        last_time = now;
    }

    return NULL;
}

int main(int argc, char **argv)
{
    int duration_sec = DEFAULT_DURATION_SEC;
    int num_readers = DEFAULT_READER_THREADS;
    int num_writers = DEFAULT_WRITER_THREADS;

    if (argc > 1) duration_sec = atoi(argv[1]);
    if (argc > 2) num_readers = atoi(argv[2]);
    if (argc > 3) num_writers = atoi(argv[3]);

    /* Enable locale for number formatting */
    setlocale(LC_NUMERIC, "");

    printf("Sauron Concurrent Stress Test\n");
    printf("==============================\n");
    printf("Duration: %d seconds\n", duration_sec);
    printf("Reader threads: %d\n", num_readers);
    printf("Writer threads: %d\n", num_writers);
    printf("Bulk loader threads: 1\n");
    printf("Decay thread: 1 (every %dms)\n", DECAY_INTERVAL_MS);
    printf("Library version: %s\n\n", sauron_version());

    /* Create context */
    g_ctx = sauron_create();
    if (!g_ctx) {
        fprintf(stderr, "ERROR: Failed to create context\n");
        return 1;
    }

    /* Pre-populate with some data */
    printf("Pre-populating with 100K entries...\n");
    uint32_t rng = 12345;
    for (int i = 0; i < 100000; i++) {
        uint32_t ip = random_valid_ip(&rng);
        sauron_set_u32(g_ctx, ip, (int16_t)(i % 1000));
    }
    printf("Initial count: %'lu, blocks: %lu\n\n",
           (unsigned long)sauron_count(g_ctx),
           (unsigned long)sauron_block_count(g_ctx));

    /* Initialize global state */
    atomic_store(&g_running, 1);
    atomic_store(&g_total_reads, 0);
    atomic_store(&g_total_writes, 0);
    atomic_store(&g_total_decays, 0);
    atomic_store(&g_read_errors, 0);
    atomic_store(&g_write_errors, 0);

    /* Allocate thread handles */
    int total_threads = num_readers + num_writers + 3;  /* +bulk +decay +stats */
    pthread_t *threads = malloc(total_threads * sizeof(pthread_t));

    printf("Starting concurrent operations...\n\n");
    double start_time = get_time();

    int t = 0;

    /* Start reader threads */
    for (int i = 0; i < num_readers; i++) {
        pthread_create(&threads[t++], NULL, reader_thread, NULL);
    }

    /* Start writer threads */
    for (int i = 0; i < num_writers; i++) {
        pthread_create(&threads[t++], NULL, writer_thread, NULL);
    }

    /* Start bulk loader thread */
    pthread_create(&threads[t++], NULL, bulk_loader_thread, NULL);

    /* Start decay thread */
    pthread_create(&threads[t++], NULL, decay_thread, NULL);

    /* Start stats thread */
    pthread_create(&threads[t++], NULL, stats_thread, NULL);

    /* Run for specified duration */
    sleep(duration_sec);

    /* Signal threads to stop */
    atomic_store(&g_running, 0);

    printf("\nStopping threads...\n");

    /* Wait for all threads */
    for (int i = 0; i < total_threads; i++) {
        pthread_join(threads[i], NULL);
    }

    double elapsed = get_time() - start_time;

    /* Get final results */
    uint64_t total_reads = atomic_load(&g_total_reads);
    uint64_t total_writes = atomic_load(&g_total_writes);
    uint64_t total_decays = atomic_load(&g_total_decays);
    uint64_t read_errors = atomic_load(&g_read_errors);
    uint64_t write_errors = atomic_load(&g_write_errors);

    printf("\n==============================\n");
    printf("RESULTS\n");
    printf("==============================\n");
    printf("Duration: %.2f seconds\n", elapsed);
    printf("Total reads: %'lu (%.2fM/sec)\n",
           (unsigned long)total_reads, total_reads / elapsed / 1e6);
    printf("Total writes: %'lu (%.2fM/sec)\n",
           (unsigned long)total_writes, total_writes / elapsed / 1e6);
    printf("Total decay modifications: %'lu\n", (unsigned long)total_decays);
    printf("Read errors: %lu\n", (unsigned long)read_errors);
    printf("Write errors: %lu\n", (unsigned long)write_errors);
    printf("Final count: %'lu\n", (unsigned long)sauron_count(g_ctx));
    printf("Final blocks: %lu\n", (unsigned long)sauron_block_count(g_ctx));
    printf("Final memory: %.2f MB\n", sauron_memory_usage(g_ctx) / (1024.0 * 1024.0));

    int pass = (read_errors == 0 && write_errors == 0);
    printf("\nRESULT: %s\n", pass ? "PASS - No errors detected" : "FAIL - Errors detected");

    /* Cleanup */
    sauron_destroy(g_ctx);
    free(threads);

    return pass ? 0 : 1;
}

File: examples/example_concurrent.py
#!/usr/bin/env python3
"""
Sauron Library - Concurrent Stress Test (Python)

Exercises thread safety by running simultaneous:
- Random IP readers (multiple threads)
- Bulk score updaters (multiple threads)
- Decay operations (single thread, periodic)

Note: Python's GIL means threads don't run truly in parallel for Python code,
but the C library operations do release the GIL during execution, allowing
real concurrency at the C level.

Run:
    LD_LIBRARY_PATH=../src/.libs python3 example_concurrent.py [duration] [readers] [writers]
    LD_LIBRARY_PATH=../src/.libs python3 example_concurrent.py 10 4 2

Copyright (c) 2024-2026, Ron Dilley
"""

import os
import sys
import time
import random
import threading
import argparse
from pathlib import Path
from dataclasses import dataclass
from typing import List

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent / "python"))

from sauron import Sauron, SAURON_SCORE_MIN, SAURON_SCORE_MAX


@dataclass
class ThreadStats:
    """Statistics collected by each thread."""
    reads: int = 0
    writes: int = 0
    decays: int = 0
    read_errors: int = 0
    write_errors: int = 0


class ConcurrentStressTest:
    """Concurrent stress test for Sauron library."""

    def __init__(self, duration: int, num_readers: int, num_writers: int):
        self.duration = duration
        self.num_readers = num_readers
        self.num_writers = num_writers
        self.sauron = Sauron()
        self.running = threading.Event()
        self.running.set()
        self.stats_lock = threading.Lock()
        self.all_stats: List[ThreadStats] = []

    def random_valid_ip(self) -> int:
        """Generate a random non-bogon IP as uint32."""
        while True:
            # Generate random IP avoiding known bogon ranges
            first_octet = random.randint(1, 223)
            if first_octet == 127:
                continue  # Skip loopback

            ip = (first_octet << 24) | random.randint(0, 0xFFFFFF)

            if not self.sauron.is_bogon(ip):
                return ip

    def reader_thread(self, thread_id: int) -> ThreadStats:
        """Reader thread: continuously reads random IPs."""
        stats = ThreadStats()
        local_rng = random.Random(thread_id + time.time_ns())

        while self.running.is_set():
            try:
                ip = self.random_valid_ip()
                score = self.sauron.get_u32(ip)

                # Validate score is in range
                if score < SAURON_SCORE_MIN or score > SAURON_SCORE_MAX:
                    stats.read_errors += 1

                stats.reads += 1

                # Yield occasionally
                if stats.reads % 10000 == 0:
                    time.sleep(0)  # Yield to other threads

            except Exception as e:
                stats.read_errors += 1

        return stats

    def writer_thread(self, thread_id: int) -> ThreadStats:
        """Writer thread: continuously sets/increments random IPs."""
        stats = ThreadStats()
        local_rng = random.Random(thread_id + time.time_ns() + 12345)

        while self.running.is_set():
            try:
                ip = self.random_valid_ip()
                value = local_rng.randint(-500, 500)
                op = local_rng.randint(0, 2)

                if op == 0:
                    result = self.sauron.set_u32(ip, value)
                elif op == 1:
                    result = self.sauron.incr_u32(ip, value)
                else:
                    result = self.sauron.decr_u32(ip, abs(value))

                # Validate result
                if result < SAURON_SCORE_MIN or result > SAURON_SCORE_MAX:
                    stats.write_errors += 1

                stats.writes += 1

                # Yield occasionally
                if stats.writes % 1000 == 0:
                    time.sleep(0)

            except Exception as e:
                stats.write_errors += 1

        return stats

    def bulk_loader_thread(self, thread_id: int) -> ThreadStats:
        """Bulk loader thread: periodically bulk loads batches."""
        stats = ThreadStats()
        batch_size = 5000

        while self.running.is_set():
            try:
                # Generate batch of updates
                lines = []
                for _ in range(batch_size):
                    ip = self.random_valid_ip()
                    value = random.randint(-100, 100)
                    is_relative = random.random() < 0.5

                    ip_str = Sauron.u32_to_ip(ip)
                    if is_relative:
                        if value >= 0:
                            lines.append(f"{ip_str},+{value}")
                        else:
                            lines.append(f"{ip_str},+{value}")  # +negative for relative decrement
                    else:
                        lines.append(f"{ip_str},{value}")

                data = "\n".join(lines)
                result = self.sauron.bulk_load_string(data)
                stats.writes += result.sets + result.updates

                # Small delay between bulk loads
                time.sleep(0.01)

            except Exception as e:
                stats.write_errors += 1

        return stats

    def decay_thread(self, thread_id: int) -> ThreadStats:
        """Decay thread: periodically applies decay."""
        stats = ThreadStats()
        decay_interval = 0.5  # seconds

        while self.running.is_set():
            try:
                modified = self.sauron.decay(0.99, 1)
                stats.decays += modified
                time.sleep(decay_interval)
            except Exception as e:
                pass  # Decay errors are non-fatal

        return stats

    def stats_reporter_thread(self):
        """Periodically reports statistics."""
        interval = 1
        last_reads = 0
        last_writes = 0
        last_time = time.time()
        elapsed_sec = 0

        while self.running.is_set():
            time.sleep(interval)
            elapsed_sec += interval

            # Aggregate current stats
            total_reads = sum(s.reads for s in self.all_stats)
            total_writes = sum(s.writes for s in self.all_stats)

            now = time.time()
            dt = now - last_time

            read_rate = (total_reads - last_reads) / dt / 1e6
            write_rate = (total_writes - last_writes) / dt / 1e6

            print(f"  [{elapsed_sec:2d}s] Reads: {total_reads:>12,} ({read_rate:.2f}M/s) | "
                  f"Writes: {total_writes:>12,} ({write_rate:.2f}M/s) | "
                  f"Count: {self.sauron.count():,} | Blocks: {self.sauron.block_count()}")

            last_reads = total_reads
            last_writes = total_writes
            last_time = now

    def run_thread_with_stats(self, target, thread_id: int):
        """Wrapper to run thread and collect stats."""
        stats = target(thread_id)
        with self.stats_lock:
            self.all_stats.append(stats)

    def run(self):
        """Run the concurrent stress test."""
        print("Sauron Concurrent Stress Test (Python)")
        print("=" * 50)
        print(f"Duration: {self.duration} seconds")
        print(f"Reader threads: {self.num_readers}")
        print(f"Writer threads: {self.num_writers}")
        print(f"Bulk loader threads: 1")
        print(f"Decay thread: 1 (every 500ms)")
        print(f"Library version: {self.sauron.version}")
        print()

        # Pre-populate with some data
        print("Pre-populating with 100K entries...")
        for i in range(100000):
            ip = self.random_valid_ip()
            self.sauron.set_u32(ip, i % 1000)

        print(f"Initial count: {self.sauron.count():,}, blocks: {self.sauron.block_count()}")
        print()

        # Create thread list
        threads = []
        self.all_stats = [ThreadStats() for _ in range(self.num_readers + self.num_writers + 2)]

        print("Starting concurrent operations...")
        print()
        start_time = time.time()

        # Start reader threads
        for i in range(self.num_readers):
            t = threading.Thread(
                target=self._reader_wrapper,
                args=(i,),
                daemon=True
            )
            threads.append(t)
            t.start()

        # Start writer threads
        for i in range(self.num_writers):
            t = threading.Thread(
                target=self._writer_wrapper,
                args=(self.num_readers + i,),
                daemon=True
            )
            threads.append(t)
            t.start()

        # Start bulk loader thread
        t = threading.Thread(
            target=self._bulk_wrapper,
            args=(self.num_readers + self.num_writers,),
            daemon=True
        )
        threads.append(t)
        t.start()

        # Start decay thread
        t = threading.Thread(
            target=self._decay_wrapper,
            args=(self.num_readers + self.num_writers + 1,),
            daemon=True
        )
        threads.append(t)
        t.start()

        # Start stats reporter
        stats_thread = threading.Thread(target=self.stats_reporter_thread, daemon=True)
        stats_thread.start()

        # Run for specified duration
        time.sleep(self.duration)

        # Signal threads to stop
        self.running.clear()
        print("\nStopping threads...")

        # Wait for threads (with timeout)
        for t in threads:
            t.join(timeout=2.0)

        elapsed = time.time() - start_time

        # Aggregate results
        total_reads = sum(s.reads for s in self.all_stats)
        total_writes = sum(s.writes for s in self.all_stats)
        total_decays = sum(s.decays for s in self.all_stats)
        read_errors = sum(s.read_errors for s in self.all_stats)
        write_errors = sum(s.write_errors for s in self.all_stats)

        print()
        print("=" * 50)
        print("RESULTS")
        print("=" * 50)
        print(f"Duration: {elapsed:.2f} seconds")
        print(f"Total reads: {total_reads:,} ({total_reads/elapsed/1e6:.2f}M/sec)")
        print(f"Total writes: {total_writes:,} ({total_writes/elapsed/1e6:.2f}M/sec)")
        print(f"Total decay modifications: {total_decays:,}")
        print(f"Read errors: {read_errors}")
        print(f"Write errors: {write_errors}")
        print(f"Final count: {self.sauron.count():,}")
        print(f"Final blocks: {self.sauron.block_count()}")
        print(f"Final memory: {self.sauron.memory_usage() / (1024*1024):.2f} MB")

        passed = read_errors == 0 and write_errors == 0
        print(f"\nRESULT: {'PASS - No errors detected' if passed else 'FAIL - Errors detected'}")

        self.sauron.close()
        return 0 if passed else 1

    def _reader_wrapper(self, idx: int):
        """Wrapper for reader thread to store stats."""
        stats = self.reader_thread(idx)
        self.all_stats[idx] = stats

    def _writer_wrapper(self, idx: int):
        """Wrapper for writer thread to store stats."""
        stats = self.writer_thread(idx)
        self.all_stats[idx] = stats

    def _bulk_wrapper(self, idx: int):
        """Wrapper for bulk loader thread to store stats."""
        stats = self.bulk_loader_thread(idx)
        self.all_stats[idx] = stats

    def _decay_wrapper(self, idx: int):
        """Wrapper for decay thread to store stats."""
        stats = self.decay_thread(idx)
        self.all_stats[idx] = stats


def main():
    parser = argparse.ArgumentParser(description="Sauron Concurrent Stress Test")
    parser.add_argument("duration", type=int, nargs="?", default=10,
                        help="Test duration in seconds (default: 10)")
    parser.add_argument("readers", type=int, nargs="?", default=4,
                        help="Number of reader threads (default: 4)")
    parser.add_argument("writers", type=int, nargs="?", default=2,
                        help="Number of writer threads (default: 2)")
    args = parser.parse_args()

    test = ConcurrentStressTest(args.duration, args.readers, args.writers)
    return test.run()


if __name__ == "__main__":
    sys.exit(main())
